Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1636,22 +1673,126 @@ void CGContextSetCMYKFillColor(CGContextRef context, CGFloat cyan, CGFloat magen
 }
 #pragma endregion
 
+#pragma region Image Helpers - CGImage / Brush
+
+static HRESULT __CreateD2DBitmapFromCGImage(CGContextRef context, CGImageRef image, ID2D1Bitmap** bitmap) {
+    RETURN_HR_IF_NULL(E_INVALIDARG, context);
+    RETURN_HR_IF_NULL(E_INVALIDARG, image);
+    RETURN_HR_IF_NULL(E_POINTER, bitmap);
+
+    ComPtr<IWICBitmap> bmap;
+    RETURN_IF_FAILED(_CGImageGetWICImageSource(image, &bmap));
+
+    ComPtr<ID2D1Bitmap> d2dBitmap;
+    RETURN_IF_FAILED(context->DeviceContext()->CreateBitmapFromWicBitmap(bmap.Get(), nullptr, &d2dBitmap));
+    *bitmap = d2dBitmap.Detach();
+    return S_OK;
+}
+
+static CGImageRef __CGContextCreateRenderableImage(CGImageRef image) {
+    RETURN_NULL_IF(!image);
+    WICPixelFormatGUID imagePixelFormat = _CGImageGetWICPixelFormat(image);
+    if (!_CGIsValidRenderTargetPixelFormat(imagePixelFormat)) {
+        // convert it to a valid pixelformat
+        return _CGImageCreateCopyWithPixelFormat(image, GUID_WICPixelFormat32bppPBGRA);
+    }
+
+    return CGImageRetain(image);
+}
+
+#pragma endregion
+
 #pragma region Drawing Parameters - Stroke / Fill Patterns
+
+template <typename ContextStageLambda> // Takes the form void(*)(CGContextRef)
+static HRESULT _CreatePatternBrush(
+    CGContextRef context, CGPatternRef pattern, const CGFloat* components, ID2D1BitmapBrush1** brush, ContextStageLambda&& contextStage) {
+    // TODO: change to support grayscale (masks) after dustins change.
+    woc::unique_cf<CGColorSpaceRef> colorspace{ CGColorSpaceCreateDeviceRGB() };
+
+    // We need to generate the pattern as an image (then tile it)
+    CGRect tileSize = _CGPatternGetFinalPatternSize(pattern);
+    size_t bitsPerComponent = 8;
+    size_t bytesPerRow = 4 * tileSize.size.width;
+    CGBitmapInfo bitmapInfo = kCGImageAlphaPremultipliedFirst | kCGBitmapByteOrder32Big;
+
+    woc::unique_cf<CGContextRef> patternContext{ CGBitmapContextCreate(
+        nullptr, tileSize.size.width, tileSize.size.height, bitsPerComponent, bytesPerRow, colorspace.get(), bitmapInfo) };
+    RETURN_HR_IF_NULL(E_UNEXPECTED, patternContext);
+
+    // Stage the drawing context
+    RETURN_IF_FAILED(std::forward<ContextStageLambda>(contextStage)(patternContext.get()));
```

Comment:
the lamda returns HRESULT LOL

---
In reply to: [93076852](https://github.com/Microsoft/WinObjC/pull/1586#discussion_r93076852) [](ancestors = 93076852)